src/mon/MonMap.h

```diff
@@ -25,34 +25,124 @@ namespace ceph {
   class Formatter;
 }
 
+struct mon_info_t {
+  /**
+   * monitor name
+   *
+   * i.e., 'foo' in 'mon.foo'
+   */
+  string name;
+  /**
+   * monitor's public address
+   *
+   * public facing address, traditionally used to communicate with all clients
+   * and other monitors.
+   */
+  entity_addr_t public_addr;
+
+  mon_info_t(string &n, entity_addr_t& p_addr)
+    : name(n), public_addr(p_addr)
+  { }
+
+  mon_info_t() { }
+
+
+  void encode(bufferlist& bl) const;
+  void decode(bufferlist::iterator& p);
+  void print(ostream& out) const;
+};
+WRITE_CLASS_ENCODER(mon_info_t)
+inline ostream& operator<<(ostream& out, const mon_info_t& mon) {
+  mon.print(out);
+  return out;
+}
+
+
 class MonMap {
  public:
   epoch_t epoch;       // what epoch/version of the monmap
   uuid_d fsid;
-  map<string, entity_addr_t> mon_addr;
   utime_t last_changed;
   utime_t created;
 
-  map<entity_addr_t,string> addr_name;
+  /* we don't use this anymore, but we keep it around to make
+   * compatibility easier, as there may be monitors that do not
+   * support our way to do monmaps (e.g., during an upgrade).
+   *
+   * This map is kept populated at all times with the public address of
+   * each entry in 'mons'.
+   *
+   * This will be encoded with every map we produce just as well. We could
+   * generate this during encoding, but seems like that's just wasting
+   * cycles. And this barelly wastes any resources, so there's no big
+   * tradeoff anyway.
+   */
+  map<string, entity_addr_t> mon_addr;
+
+  map<string, mon_info_t> mons;
+  map<entity_addr_t, mon_info_t> addr_mons;
+
+  vector<mon_info_t> ranks;
```

Comment:
Hmm, note that once all mon understand the new rank ordering they will work properly, but old clients will call things mon.2 that are actually mon.1 or something, which might be confusing. We also should audit the MonClient and kernel mon_client.c to ensure they don't assert the received message is from who they think it's from.. that could be problematic!  I think MonClient goes but Connection*, so that's fine, but the kernel might go by the from int and get confused.